### PR TITLE
rocmPackages.rocprofiler: fix rocprof and rocprofv2

### LINKED ISSUE #262615: rocprofv2: aqlprofile API table load failed
### Describe the bug
When trying to use rocprofv2:
```
❯ rocprofv2 ./a.out 10
aqlprofile API table load failed: HSA_STATUS_ERROR: A generic error has occurred.
/nix/store/c5iy8c6w34y4jfgqlv8c6scxrilc4djv-rocprofiler-5.7.0/bin/rocprofv2: line 264: 77294 Aborted                 (core dumped) LD_PRELOAD=$LD_PRELOAD:$ROCM_DIR/lib/rocprofiler/librocprofiler_tool.so $*
```
This might have something to do with some predefined paths in [rocprofv2](https://github.com/ROCm-Developer-Tools/rocprofiler/blob/11afaaee12fcd4a3927e6afaeda1216d415b958b/bin/rocprofv2#L263)

### Steps To Reproduce
1. Compile an executable with hipcc
2. Open it with rocprofv2

### Notify maintainers

 @NixOS/rocm-maintainers 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.7, NixOS, 23.11 (Tapir), 23.11.20231016.ca012a0`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - nixpkgs: `/run/current-system/nixpkgs`
```

### PR DESCRIPTION
**Pull request by Madouura**
## Description of changes
Tracking: #197885
Fixes: #262615

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/rocm-modules/5/rocprofiler/default.nix [hunk 1]
```diff
@@ -79,16 +93,31 @@ stdenv.mkDerivation (finalAttrs: {
...
+    substituteInPlace test/CMakeLists.txt \
+      --replace "\''${ROCM_ROOT_DIR}/amdgcn/bitcode" "${rocm-device-libs}/amdgcn/bitcode"
   '';
 
-  postBuild = ''
-    # HSACO aren't being built for some reason
-    substituteInPlace test/cmake_install.cmake \
-      --replace "file(INSTALL DESTINATION \"\''${CMAKE_INSTALL_PREFIX}/share/rocprofiler/tests-v1\" TYPE FILE FILES \"" "message(\""
+  postInstall = ''
+    # ???
```
Comment (SuperSandro2000) [inline pkgs/development/rocm-modules/5/rocprofiler/default.nix:118]: I am not sure what that comment should tell the reader

File: pkgs/development/rocm-modules/5/rocprofiler/default.nix [hunk 2]
```diff
@@ -79,16 +93,31 @@ stdenv.mkDerivation (finalAttrs: {
   postPatch = ''
     patchShebangs .
```
Comment (SuperSandro2000) [inline pkgs/development/rocm-modules/5/rocprofiler/default.nix:94]: not sure how feasible it is to narrow this down

File: pkgs/development/rocm-modules/5/rocprofiler/default.nix [hunk 3]
```diff
@@ -79,16 +93,31 @@ stdenv.mkDerivation (finalAttrs: {
...
+    substituteInPlace bin/rocprofv2 \
+      --replace "ROCM_DIR=\$(dirname -- \"\$ROCPROFV2_DIR\")" "ROCPROFILER_DIR=\$(dirname -- \"\$ROCPROFV2_DIR\")''\nROCM_DIR=${rocmtoolkit-merged}" \
+      --replace "\$ROCM_DIR/lib/rocprofiler" "\$ROCPROFILER_DIR/lib/rocprofiler" \
+      --replace "\$ROCM_DIR/libexec/rocprofiler" "\$ROCPROFILER_DIR/libexec/rocprofiler"
+
+    substituteInPlace tests-v2/featuretests/profiler/CMakeLists.txt \
+      --replace "--build-id=sha1" "--build-id=sha1 --rocm-path=${clr} --rocm-device-lib-path=${rocm-device-libs}/amdgcn/bitcode"
+
+    substituteInPlace test/CMakeLists.txt \
+      --replace "\''${ROCM_ROOT_DIR}/amdgcn/bitcode" "${rocm-device-libs}/amdgcn/bitcode"
```
Comment (SuperSandro2000) [inline pkgs/development/rocm-modules/5/rocprofiler/default.nix:114]: not sure if a patch would be easier to manage

File: pkgs/development/rocm-modules/5/rocprofiler/default.nix (hunk 1 above)
Comment (Madouura) [reply pkgs/development/rocm-modules/5/rocprofiler/default.nix:118]: Confusion, on why the two libraries below didn't already have the executable bit.

File: pkgs/development/rocm-modules/5/rocprofiler/default.nix (hunk 2 above)
Comment (Madouura) [reply pkgs/development/rocm-modules/5/rocprofiler/default.nix:94]: There's like two directories with a few files that we can cut out.
IMO not worth narrowing down.

File: pkgs/development/rocm-modules/5/rocprofiler/default.nix (hunk 3 above)
Comment (Madouura) [reply pkgs/development/rocm-modules/5/rocprofiler/default.nix:114]: I'm hesitant to use a patch for the shell scripts.

File: pkgs/development/rocm-modules/5/rocprofiler/default.nix (hunk 2 above)
Comment (Madouura) [reply pkgs/development/rocm-modules/5/rocprofiler/default.nix:94]: Actually scratch that, it's literally only ``build.sh`` that we could cut out.
Everything else should probably stay.

**Comment by pbsds** [comment]
these seem to be marked executable:

```
results/rocmPackages.rocprofiler/libexec/rocprofiler/counters/basic_counters.xml
results/rocmPackages.rocprofiler/libexec/rocprofiler/counters/derived_counters.xml
```

**Comment by Madouura** [comment]
So many weird things going on with this program.
Anyway thank you for pointing that out!

**Comment by pbsds** [comment]
I'd like to learn more about the rocm tooling, how may I test this? I have a 580 if it helps

**Comment by Madouura** [comment]
A basic test for ``rocprofiler`` is pretty easy.
See: https://github.com/NixOS/nixpkgs/issues/262615#issuecomment-1774043231
1. Compile it with ``hipcc`` (``rocmPackages.clr``)
2. Make sure it runs ``./a.out``
3. Run it with rocprofiler ``rocprof --hip-trace a.out`` & ``rocprofv2 a.out``
``rocprof`` needs a ``--`` option for any useful output. (``--stats`` ``--roctx-trace`` ``--hip-trace`` ``--hsa-trace`` ``--sys-trace``)
I suggest just using ``rocprofv2`` instead anyway.

**Comment by Madouura** [comment]
That said, I'm not entirely sure anything you do with ``rocprofiler`` will succeed on a 580.
I don't think polaris is supported in ROCm?

**Comment by pbsds** [comment]
Been at it for a while now, but it seems polaris support is truly dropped. Diff LGTM

**Review by pbsds — APPROVED** [review]

**Comment by Flakebi** [comment]
You should be able to run rocm (or at least part of it) on polaris when you set the environment variable `ROC_ENABLE_PRE_VEGA=1` (at runtime).

**Comment by pbsds** [comment]
Been trying it out, rebuilding my system with rocm stuff enabled, and no luck... My repro for those interested:
```shell
NIXPKGS_ALLOW_UNFREE=1 ROC_ENABLE_PRE_VEGA=1 nix shell --impure github:Madouura/nixpkgs/pr/rocprofiler#rocmPackages.{clr,rocprofiler} -c bash -x <(cat <<'EOF'
    cd "$(mktemp -d)"

    curl -L https://gist.githubusercontent.com/pbsds/aab148188bdcf308f70053d3e53cefb4/raw/935bfe11709e9e3b69b89435bcd487c69a742e5b/test.hip > test.hip
    hipcc test.hip
    ./a.out
    rocprof --hip-trace ./a.out
    rocprofv2 ./a.out

    rm -rf "$(pwd)"
EOF
)
```


output: 

<details>

```
+ hipcc test.hip
+ ./a.out
DEVICE: 
Max thread per block: 0
/dev/fd/63: line 5: 21052 Floating point exception(core dumped) ./a.out
+ rocprof --hip-trace ./a.out
RPL: on '231026_160503' from '/nix/store/2sml94akzdgpm9vvdbpxf096d2gcnnhp-rocprofiler-5.7.1' in '/tmp/tmp.bZ4F6eKy5e'
RPL: profiling '"./a.out"'
RPL: input file ''
RPL: output dir '/tmp/rpl_data_231026_160503_21061'
RPL: result dir '/tmp/rpl_data_231026_160503_21061/input_results_231026_160503'
ROCtracer (21083):
    HIP-trace(*)
/nix/store/2sml94akzdgpm9vvdbpxf096d2gcnnhp-rocprofiler-5.7.1/bin/rocprof: line 301: 21083 Floating point exception(core dumped) "./a.out"
+ rocprofv2 ./a.out
DEVICE: hg���
Max thread per block: 0
/nix/store/2sml94akzdgpm9vvdbpxf096d2gcnnhp-rocprofiler-5.7.1/bin/rocprofv2: line 247: 21138 Floating point exception(core dumped) LD_PRELOAD=$LD_PRELOAD:$ROCPROFILER_DIR/lib/rocprofiler/librocprofiler_tool.so $*
```
</details>

Could you post your result @flakebi ? I'm sorry if I'm being overly cautious.

**Comment by Flakebi** [comment]
I get this on a gfx10 / navi10 card (RX 5700 XT):
```
++ mktemp -d
+ cd /run/user/1000/tmp.p4asAXbx5a
+ curl -L https://gist.githubusercontent.com/pbsds/aab148188bdcf308f70053d3e53cefb4/raw/935bfe11709e9e3b69b89435bcd487c69a742e5b/test.hip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   977  100   977    0     0   3523      0 --:--:-- --:--:-- --:--:--  3527
+ hipcc test.hip
+ ./a.out
DEVICE: AMD Radeon RX 5700 XT
Max thread per block: 1024
0
1
2
3
4
5
6
7
8
9
+ rocprof --hip-trace ./a.out
RPL: on '231026_163017' from '/nix/store/h7piknhdjnv4z4y880dlg2xbd3c9x665-rocprofiler-5.7.1' in '/run/user/1000/tmp.p4asAXbx5a'
RPL: profiling '"./a.out"'
RPL: input file ''
RPL: output dir '/tmp/rpl_data_231026_163017_3895426'
/nix/store/h7piknhdjnv4z4y880dlg2xbd3c9x665-rocprofiler-5.7.1/bin/rocprof: line 560: /bin/ls: No such file or directory
Traceback (most recent call last):
  File "/nix/store/h7piknhdjnv4z4y880dlg2xbd3c9x665-rocprofiler-5.7.1/libexec/rocprofiler/tblextr.py", line 797, in <module>
    indir = re.sub(r'\/[^\/]*$', r'', infiles[0])
                                      ~~~~~~~^^^
IndexError: list index out of range
Profiling data corrupted: ''
+ rocprofv2 ./a.out
aqlprofile API table load failed: HSA_STATUS_ERROR: A generic error has occurred.
/nix/store/h7piknhdjnv4z4y880dlg2xbd3c9x665-rocprofiler-5.7.1/bin/rocprofv2: line 246: 3895453 Aborted                 (core dumped) LD_PRELOAD=$LD_PRELOAD:$ROCM_DIR/lib/rocprofiler/librocprofiler_tool.so $*
++ pwd
+ rm -rf /run/user/1000/tmp.p4asAXbx5a
```

**Comment by Madouura** [comment]
Only things that don't work for me is ``rocprof`` with ``--hsa-trace``, and the default ``--stats``.
I can't get ``rocprofv2`` to fail at all.

**Comment by Madouura** [comment]
> I get this on a gfx10 / navi10 card (RX 5700 XT):
> 
> ```
> ++ mktemp -d
> + cd /run/user/1000/tmp.p4asAXbx5a
> + curl -L https://gist.githubusercontent.com/pbsds/aab148188bdcf308f70053d3e53cefb4/raw/935bfe11709e9e3b69b89435bcd487c69a742e5b/test.hip
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100   977  100   977    0     0   3523      0 --:--:-- --:--:-- --:--:--  3527
> + hipcc test.hip
> + ./a.out
> DEVICE: AMD Radeon RX 5700 XT
> Max thread per block: 1024
> 0
> 1
> 2
> 3
> 4
> 5
> 6
> 7
> 8
> 9
> + rocprof --hip-trace ./a.out
> RPL: on '231026_163017' from '/nix/store/h7piknhdjnv4z4y880dlg2xbd3c9x665-rocprofiler-5.7.1' in '/run/user/1000/tmp.p4asAXbx5a'
> RPL: profiling '"./a.out"'
> RPL: input file ''
> RPL: output dir '/tmp/rpl_data_231026_163017_3895426'
> /nix/store/h7piknhdjnv4z4y880dlg2xbd3c9x665-rocprofiler-5.7.1/bin/rocprof: line 560: /bin/ls: No such file or directory
> Traceback (most recent call last):
>   File "/nix/store/h7piknhdjnv4z4y880dlg2xbd3c9x665-rocprofiler-5.7.1/libexec/rocprofiler/tblextr.py", line 797, in <module>
>     indir = re.sub(r'\/[^\/]*$', r'', infiles[0])
>                                       ~~~~~~~^^^
> IndexError: list index out of range
> Profiling data corrupted: ''
> + rocprofv2 ./a.out
> aqlprofile API table load failed: HSA_STATUS_ERROR: A generic error has occurred.
> /nix/store/h7piknhdjnv4z4y880dlg2xbd3c9x665-rocprofiler-5.7.1/bin/rocprofv2: line 246: 3895453 Aborted                 (core dumped) LD_PRELOAD=$LD_PRELOAD:$ROCM_DIR/lib/rocprofiler/librocprofiler_tool.so $*
> ++ pwd
> + rm -rf /run/user/1000/tmp.p4asAXbx5a
> ```

Noticed the /bin/ls error.
I don't think you're using this PR's patch, since that shouldn't be coming up.

**Comment by Flakebi** [comment]
> I don't think you're using this PR's patch, since that shouldn't be coming up.

Yeah, I merged this branch into some other branch, I would have expected it to work. Anyway, I tried again without merging and it seems to work fine.

<details>
<summary>Output</summary>

```
++ mktemp -d
+ cd /tmp/tmp.ulzeX6HqGV
+ curl -L https://gist.githubusercontent.com/pbsds/aab148188bdcf308f70053d3e53cefb4/raw/935bfe11709e9e3b69b89435bcd487c69a742e5b/test.hip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   977  100   977    0     0   2612      0 --:--:-- --:--:-- --:--:--  2612
+ hipcc test.hip
+ ./a.out
DEVICE: AMD Radeon RX 5700 XT
Max thread per block: 1024
0
1
2
3
4
5
6
7
8
9
+ rocprof --hip-trace ./a.out
RPL: on '231027_123732' from '/nix/store/2sml94akzdgpm9vvdbpxf096d2gcnnhp-rocprofiler-5.7.1' in '/tmp/tmp.ulzeX6HqGV'
RPL: profiling '"./a.out"'
RPL: input file ''
RPL: output dir '/tmp/rpl_data_231027_123732_4079854'
RPL: result dir '/tmp/rpl_data_231027_123732_4079854/input_results_231027_123732'
ROCtracer (4079874):
    HIP-trace(*)
DEVICE: AMD Radeon RX 5700 XT
Max thread per block: 1024
0
1
2
3
4
5
6
7
8
9
hsa_copy_deps: 0
scan ops data 1:2                                                                                                    File '/tmp/tmp.ulzeX6HqGV/results.copy_stats.csv' is generating
dump json 0:1
File '/tmp/tmp.ulzeX6HqGV/results.json' is generating
File '/tmp/tmp.ulzeX6HqGV/results.hip_stats.csv' is generating
dump json 11:12
File '/tmp/tmp.ulzeX6HqGV/results.json' is generating
File '/tmp/tmp.ulzeX6HqGV/results.stats.csv' is generating
dump json 0:1
File '/tmp/tmp.ulzeX6HqGV/results.json' is generating
+ rocprofv2 ./a.out
DEVICE: AMD Radeon RX 5700 XT
Max thread per block: 1024
0
1
2
3
4
5
6
7
8
9
Dispatch_ID(1), GPU_ID(1), Queue_ID(1), Queue_Index(0), Process_ID(4079928), Thread_ID(4079928), Grid_Size(1024), Workgroup_Size(1024), LDS(0), Scratch_Size(0), Arch_VGPR(8), Accumulative_VGPR(0), SGPR(128), Wave_Size(32), Kernel_Name("gene"), Begin_Timestamp(1465566548148299), End_Timestamp(1465566548153059), Correlation_ID(0)
++ pwd
+ rm -rf /tmp/tmp.ulzeX6HqGV
```

</details>

**Comment by pbsds** [comment]
Seems like the linked issue is fixed

**Pull request merged**